File: dynamice programming/1639.js
/**
 * @param {string[]} words
 * @param {string} target
 * @return {number}
 */
const numWays = function(words, target) {
    const m = target.length;
    const k = words[0].length;
    const MOD = 1e9 + 7;
    
    // Initialize frequency array (26 letters x k columns)
    const freq = Array.from({ length: 26 }, () => new Array(k).fill(0));
    
    // Populate frequency array
    for (let col = 0; col < k; col++) {
        for (const word of words) {
            const char = word.charCodeAt(col) - 97; // 'a' = 97
            freq[char][col]++;
        }
    }
    
    // Initialize memoization table (m rows x k columns)
    const memo = Array.from({ length: m }, () => new Array(k).fill(-1));
    
    // Recursive DP function with memoization
    function dp(i, j) {
        if (i === m) return 1;  // Target fully formed
        if (j === k) return 0;  // No more columns to choose from
        
        if (memo[i][j] !== -1) return memo[i][j];
        
        // Option 1: Skip current column
        const skip = dp(i, j + 1) % MOD;
        
        // Option 2: Take current character
        const charCode = target.charCodeAt(i) - 97;
        const take = (freq[charCode][j] * dp(i + 1, j + 1)) % MOD;
        
        return memo[i][j] = (skip + take) % MOD;
    }
    
    return dp(0, 0);
};

File: dynamice programming/1639.py
# Google, Amazon

# Number of ways to form a string
# Same index cannot be used again in another word suppose if index is 1 then 0, 1 of each word index cannot use again..left to used index will be forbidden to use again for other words too

# T.C : O(n*k + m*k)
# S.C : O(m*k)

from typing import List
class Solution:
    def numWays(self, words: List[str], target: str) -> int:
        m = len(target)
        k = len(words[0])
        MOD = 10**9 + 7
        
        # Frequency array
        freq = [[0] * k for _ in range(26)]
        
        # Fill frequency array
        for col in range(k):
            for word in words:
                ch = word[col]
                freq[ord(ch) - ord('a')][col] += 1
        
        # Memoization table
        t = [[-1] * (k + 1) for _ in range(m + 1)]
        
        def solve(i, j):
            if i == m:
                return 1
            if j == k:
                return 0
            if t[i][j] != -1:
                return t[i][j]
            
            not_taken = solve(i, j + 1) % MOD
            taken = (freq[ord(target[i]) - ord('a')][j] * solve(i + 1, j + 1)) % MOD
            
            t[i][j] = (not_taken + taken) % MOD
            return t[i][j]
        
        return solve(0, 0)

